### ISSUE TITLE
Bug - Wazuh agent not re-resolving Wazuh workers DNS - agent tries to reach a dead IP

### ISSUE THREAD

**Issue by JPLachance** [wazuh/wazuh#4361]
| 3.10.2-1 | ossec-agentd | Agent | Debian package (using the [Wazuh Puppet module](https://github.com/wazuh/wazuh-puppet)) | Ubuntu 16.04 |

Greetings,

My Wazuh workers are deployed in Kubernetes and are behind a load balancer. In my `ossec.conf`file, the `<server>` `<address>` is an AWS Route 53 record.
```XML
<ossec_config>
  <client>
    <server>
      <address>wazuh-worker.prod.domain.com</address>
      <protocol>tcp</protocol>
      <port>1514</port>
    </server>
    <notify_time>10</notify_time>
    <time-reconnect>60</time-reconnect>
    <crypto_method>aes</crypto_method>
    <auto_restart>yes</auto_restart>
  </client>
  ...
</ossec_config>
```

It all works well, but is looks like the Wazuh agent is resolving the Worker IP address once then uses it forever. With the Kubernetes cluster autoscaling policy, Workers are moving from one node to another and the AWS Load Balancer itself never offers stable IP addresses.

So, after a couple of days, agents starts to disconnect and they fail to re-connect. From the `ossec.log` file, we read:
```
root@some-host:~# tail -10 /var/ossec/logs/ossec.log 
2019/12/11 22:09:01 ossec-agentd: ERROR: (1216): Unable to connect to '10.1.26.108': 'Connection timed out'.
2019/12/11 22:09:11 ossec-agentd: INFO: Trying to connect to server (wazuh-worker.prod.domain.com/10.1.26.108:1514/tcp).
2019/12/11 22:11:20 ossec-agentd: ERROR: (1216): Unable to connect to '10.1.26.108': 'Connection timed out'.
2019/12/11 22:11:30 ossec-agentd: INFO: Trying to connect to server (wazuh-worker.prod.domain.com/10.1.26.108:1514/tcp).
2019/12/11 22:13:40 ossec-agentd: ERROR: (1216): Unable to connect to '10.1.26.108': 'Connection timed out'.
2019/12/11 22:13:50 ossec-agentd: INFO: Trying to connect to server (wazuh-worker.prod.domain.com/10.1.26.108:1514/tcp).
2019/12/11 22:15:59 ossec-agentd: ERROR: (1216): Unable to connect to '10.1.26.108': 'Connection timed out'.
2019/12/11 22:16:09 ossec-agentd: INFO: Trying to connect to server (wazuh-worker.prod.domain.com/10.1.26.108:1514/tcp).
2019/12/11 22:18:18 ossec-agentd: ERROR: (1216): Unable to connect to '10.1.26.108': 'Connection timed out'.
2019/12/11 22:18:28 ossec-agentd: INFO: Trying to connect to server (wazuh-worker.prod.domain.com/10.1.26.108:1514/tcp).
```

When we try to resolve `wazuh-worker.prod.domain.com`, we get:
```
root@some-host:~# host wazuh-worker.prod.domain.com
wazuh-worker.prod.domain.com has address 10.1.16.127
wazuh-worker.prod.domain.com has address 10.1.26.129
wazuh-worker.prod.domain.com has address 10.1.22.73
```

So, there you have it! The Wazuh agent gets the IP then never asks the DNS server opinions again.

We had this issue with Wazuh 3.6.1 and we still have it in 3.10.2. Restarting the Wazuh agent does the trick, but doing that manual intervention on all our production servers as they disconnect does not scale!! 😄 

The error from our logs comes from [that line of code](https://github.com/wazuh/wazuh/blob/f3fcc09f1a75f51f445d7d5e96faa95d486edd57/src/client-agent/start_agent.c#L98), so my guess is that the issue comes from somewhere around [there](https://github.com/wazuh/wazuh/blob/f3fcc09f1a75f51f445d7d5e96faa95d486edd57/src/client-agent/start_agent.c#L52).

Thanks for the help!

**Comment by vikman90** [comment]
Hi @JPLachance,

I'm sorry you're having this problem. This is a known bug:

- Related issue: #2770 
- Proposed fix: #4241 

In short words, that's a design issue: the UNIX version of the agent performs _chroot_ jailing. I guess this was done for security reasons, but this isolates the agent daemon from the system and prevents the agent from resolving hostnames.

The agent can only resolve hostnames at startup. Just the code following the line you mentioned reads:
```c
if (!isChroot()) {
    resolveHostname(&agt->server[rc].rip, 5);
}
```

That means the agent will only attempt to resolve the hostname if _chroot_ is disabled —for instance, on Windows—. The PR above will definitely fix this problem in upcoming versions.

Thank you once more for such a greatly descriptive comment. I appreciate your feedback.

Please let me know if this answers your question, and consider closing this issue as we're already working on #2770.

Best regards.

**Comment by JPLachance** [comment]
Hello @vikman90,

The issue #2770 talks about a manager that "is not initially reachable". In my case, the agent was working then stopped working (disconnected).

At least, with the additional details in my issue, the GitHub search will help Wazuh users understand why their agents are disconnecting. I looked for an hour and decided to file this issue.

Also, can we expect a fix in Wazuh 3.10.3 ?

Thanks for your help!

**Comment by Zenidd** [comment]
Hello @JPLachance 

> The issue #2770 talks about a manager that "is not initially reachable". In my case, the agent was working then stopped working (disconnected).

While the use cases are slightly different, we want to implement a mechanism to allow the agents to re-resolve the domain name whenever necessary. 


> Also, can we expect a fix in Wazuh 3.10.3 ?

While this is a relevant issue, we are currently working on other also important tasks. When our workflow allow us we will introduce this issue in the work pipeline.



Greetings, JP Sáez

**Comment by Zenidd** [comment]
Hello @JPLachance 

> The issue #2770 talks about a manager that "is not initially reachable". In my case, the agent was working then stopped working (disconnected).

While the use cases are slightly different, we aim to implement a mechanism to allow every agent to re-resolve the domain name whenever necessary. That's why we relate your issue with the `is not initially reachable` one.

> Also, can we expect a fix in Wazuh 3.10.3 ?

This is a really relevant issue and we are currently working on other also important tasks so, when our workflow allows us, we will introduce this issue in the work pipeline. I think it's more likely that the fix for this problem will come in one of the upcoming major versions (3.11, 3.12..)

> At least, with the additional details in my issue, the GitHub search will help Wazuh users understand why their agents are disconnecting. I looked for an hour and decided to file this issue.

I do agree with you. Thanks for documenting here the bug so other users and Wazuh team members will save that hour if they need information about this issue. 
 

Sorry for the inconveniences and Greetings, JP Sáez

**Comment by Zenidd** [comment]
Hello @JPLachance 

> The issue #2770 talks about a manager that "is not initially reachable". In my case, the agent was working then stopped working (disconnected).

While the use cases are slightly different, we aim to implement a mechanism to allow every agent to re-resolve the domain name whenever necessary. That's why we relate your issue with the `is not initially reachable` one.

One possible solution is the one mentioned by @vikman90 in https://github.com/wazuh/wazuh/issues/2770: introducing a new module able to solve DNS's on demand for the other Wazuh daemons.

> Also, can we expect a fix in Wazuh 3.10.3 ?

This is a really relevant issue and we are currently working on other also important tasks so, when our workflow allows us we will introduce this issue in the work pipeline. I think it's more likely that the fix for this issue will come in one of the upcoming major versions (3.11, 3.12..)

> At least, with the additional details in my issue, the GitHub search will help Wazuh users understand why their agents are disconnecting. I looked for an hour and decided to file this issue.

I do agree with you. Thanks for documenting here the bug so other users and Wazuh team members will save that hour if they need information about this issue. 
 
Sorry for the inconveniences and Greetings, JP Sáez

**Comment by JPLachance** [comment]
@Zenidd,

Do you have a solution I could put un place today? On the Wazuh documentation, it is recommended to use a load balancer, but from what I'm reading, an AWS Load Balancer was never supported? I basically need an Nginx instance with an IP that won't ever change for Wazuh to work properly?

@jm404, how were you able to make Wazuh work in Kubernetes?

Thanks!

**Comment by jm404** [comment]
Hi @JPLachance,

I'm assuming your architecture is the following:

`Wazuh Agent -> Route 53 -> Kubernetes Nodes (Wazuh Worker on them)`

An AWS Load Balancer is compatible with Wazuh (**Important: using TCP port is required for port 1514 and 1515**). Until issue #2770 gets fixed, an approach could be:

`Wazuh Agent -> Route 53 -> AWS Load Balancer (NLB) -> Target Group -> EC2 instances`

This way your target IP would never change and the Load Balancer will redistribute the load. We've tested it with SSM Agents on target instances that update the Target Group and hostnames, etc...

Regarding your case with the Auto Scaling Group, the quick diagram would be the following:

`Wazuh Agent -> Route 53 -> AWS Load Balancer (NLB) -> Auto Scaling Group -> EC2 Instances`

Reference: https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-load-balancer.html

This way the Agent not re-resolving the DNS should not cause disconnection and workload would be balanced by the NLB.

*This second method has not been tested yet, please be aware of that when testing it.*

Hope this helps you to find a solution, let us know about the updates/issues!

Best regards,

Jose

**Comment by vikman90** [comment]
Hi all,

Finally, this issue has been merged into branch 3.12.

@JPLachance thank you for your feedback, and @ixjosemi thanks for your help here.

Let me close this issue.

Cheers.
